<html>
<head>

  <meta HTTP-EQUIV =" Content-Type" CONTENT =" text/html; charset=iso-8859-1">

  <meta NAME =" Description" CONTENT =" Wishart Pharmaceutical Research Group -
   PlasMapper">

   <link rel=stylesheet type="text/css" href="/PlasMapper/style/PlasMapper.css" title="default PlasMapper styles" />

  <title>PlasMapper - Help</title>

</head>

<body bgcolor="#ffffff">

<table align="center"><tr><td align="center">

<table width="650" border="0" align="center" cellspacing="0" cellpadding="0" style="border-collapse: collapse">

     <tr><td height=10 align="center">&nbsp;</td></tr>

      <tr><td align="center"><h1>PlasMapper - Help</h1></td></tr>

      <tr><td><br><br></td></tr>
      <tr><td><h2>Sequence input:</h2></td></tr>
      <tr><td><img SRC="/PlasMapper/image/blueline.gif" ALT="blue line" width="644" height="7"></td></tr>
      <tr><td width=645><b>Input sequence format:</b><br>
      The PlasMapper server accepts DNA sequences in FASTA or RAW format. Sequences in FASTA format consist of a greater-than (&gt;) symbol, a single title line, and then the sequence. RAW sequences contain only the sequence itself.<br><br>
      A sample FASTA entry is the following:<br>
      <pre>&gt;My plasmid sequence
caagcccttgtctctaaattcggtgcgctggcggggtaagcctagccattctgagtgtac
gcgagtaccggcgagccgatttgacatacgcatatcgggatcgatgtgccttggccaggt
tagcgcatcactttttgcttcgcagagtgtgatgcgctgatgtcggcgggaaaccctatc
accagggtaaagaggcataaccattgttcgcgcggattatctaatgtcgggcgtttgcaa
ccttgccgacttgaagtctgtcgtagaacgtgctcacgctgtcatttggcgccaagatcg
gttcaggacattttcaccatgttgatctcttagccatctttgatttcccttcatgaaact
cccatttaaaggggagcgaacacggcaccataagggcgtttgcagttagatggcgtgtcc
aatctcccactcctcggattgaaagactcttgaaaagggccgatccatgtcggcacacct
atcacctcctaattcgagag
      </pre>
      A sample RAW entry is the following:
      <pre>
tgttcacgggggcagcaacgctaccctaaatttacatgttgaaatcaacctcgaaaaacg
aaccatcgacggggagtgccagatagtcgacaggttaggctacacagaatcggtgtttat
tgctccgttccagatatcagtgtctagtcacatcgagtccttgccggcgtatggggggac
ttcctagtttccatgacagtcttagtgcaagtcggcgttcctcgaatctggcctcaccta
ctgatgctcgtctgatctgggaccgctgcttgagtaggatgcctgacctgtgtcaaaagt
accaagatgccgttcagccatggctctgagcaaggagagggccatcaccattttgaactg
tttgcacttcgcaatacaggggctcttggtgtaagagaca
      </pre>
      </td></tr>
      <tr><td><img SRC="/PlasMapper/image/blueline.gif" ALT="blue line" width="645" height="7"></td></tr>

      <tr><td width=645><b>Choosing a sequence from PlasMapper's library of plasmids:</b><br>
      You can select a plasmid sequence from PlasMapper's collection of commonly used plasmids:<br>
      <ol>
      <li>
      Choose a commercial supplier (or "public" for publicly available plasmids), from the list adjacent to the <b>Plasmid Library</b> button.
      </li>
      <li>
      Click <b>Plasmid Library</b> to view the names of plasmids available from the chosen supplier.
      </li>
      <li>
      Click on the name of the plasmid you wish to use. The sequences will automatically be entered into the sequence text area on PlasMapper's main page.
      </li>
      </ol><br>

      </td></tr>
      <tr><td><img SRC="/PlasMapper/image/blueline.gif" ALT="blue line" width="645" height="7"></td></tr>

      <tr><td width=645><b>Modifying a plasmid sequence using PlasMapper:</b><br>
      To facilitate the insertion of genes or other sequence fragments into a plasmid, the PlasMapper interface includes a <b>(Re)Format</b> button. Clicking <b>(Re)Format</b> adds numbering and spaces to the current sequence in the sequence text area, as in the following example:<br><br>
      A sequence prior to formatting:<br>
      <pre>
caagcccttgtctctaaattcggtgcgctggcggggtaagcctagccattctgagtgtac
gcgagtaccggcgagccgatttgacatacgcatatcgggatcgatgtgccttggccaggt
tagcgcatcactttttgcttcgcagagtgtgatgcgctgatgtcggcgggaaaccctatc
accagggtaaagaggcataaccattgttcgcgcggattatctaatgtcgggcgtttgcaa
ccttgccgacttgaagtctgtcgtagaacgtgctcacgctgtcatttggcgccaagatcg
gttcaggacattttcaccatgttgatctcttagccatctttgatttcccttcatgaaact
cccatttaaaggggagcgaacacggcaccataagggcgtttgcagttagatggcgtgtcc
aatctcccactcctcggattgaaagactcttgaaaagggccgatccatgtcggcacacct
atcacctcctaattcgagag
      </pre>
      The same sequence after formatting:
      <pre>
1       caagcccttg tctctaaatt cggtgcgctg gcggggtaag cctagccatt ctgagtgtac  60
61      gcgagtaccg gcgagccgat ttgacatacg catatcggga tcgatgtgcc ttggccaggt  120
121     tagcgcatca ctttttgctt cgcagagtgt gatgcgctga tgtcggcggg aaaccctatc  180
181     accagggtaa agaggcataa ccattgttcg cgcggattat ctaatgtcgg gcgtttgcaa  240
241     ccttgccgac ttgaagtctg tcgtagaacg tgctcacgct gtcatttggc gccaagatcg  300
301     gttcaggaca ttttcaccat gttgatctct tagccatctt tgatttccct tcatgaaact  360
361     cccatttaaa ggggagcgaa cacggcacca taagggcgtt tgcagttaga tggcgtgtcc  420
421     aatctcccac tcctcggatt gaaagactct tgaaaagggc cgatccatgt cggcacacct  480
481     atcacctcct aattcgagag  500
      </pre>
      </td></tr>
      <tr><td><img SRC="/PlasMapper/image/blueline.gif" ALT="blue line" width="645" height="7"></td></tr>

      <tr><td width=645><b>Automatic sequence filtering:</b><br>
      Note that the PlasMapper server automatically removes the title from FASTA entries. Characters other than G, A, T, or C are removed from the sequence portion of FASTA entries, and from RAW entries. This filtering step is performed to prepare the input for the feature identification steps.
      </td></tr>

      <tr><td><img SRC="/PlasMapper/image/blueline.gif" ALT="blue line" width="645" height="7"></td></tr>
      <tr><td><br><br></td></tr>
      <tr><td><h2>PlasMapper's annotation system:</h2></td></tr>
      <tr><td><img SRC="/PlasMapper/image/blueline.gif" ALT="blue line" width="645" height="7"></td></tr>

      <tr><td width=645><b>PlasMapper's Feature Site Database:</b><br>
      When generating a graphical or textual map, PlasMapper first searches its Feature Site Database (FSD), to identify common plasmid features in the sequence you supply. The FSD was compiled from an extensive survey of commercially and publicly available plasmids. The entries in the FSD have been grouped into ten different categories. The below table indicates the number of features in each category.<br><br>

      <table width="500" border="1" align="center" cellspacing="0" cellpadding="0">
      <tr><td><b>Category</b></td><td><b># of Features</b></td></tr>
      <tr><td>Promoters</td><td>51</td></tr>
      <tr><td>Regulatory sequences</td><td>63</td></tr>
      <tr><td>Terminators</td><td>29</td></tr>
      <tr><td>Selectable markers</td><td>27</td></tr>
      <tr><td>Replication origins</td><td>14</td></tr>
      <tr><td>Reporter genes</td><td>27</td></tr>
      <tr><td>Affinity tags</td><td>26</td></tr>
      <tr><td>Localization sequences</td><td>17</td></tr>
      <tr><td>Two-hybrid genes</td><td>10</td></tr>
      <tr><td>Miscellaneous features</td><td>94</td></tr>
      <table><br>
      </td></tr>
      <tr><td><img SRC="/PlasMapper/image/blueline.gif" ALT="blue line" width="645" height="7"></td></tr>

      <tr><td width=645><b>Controlling which feature types are shown:</b><br>
      To specify which feature types are shown on graphical and textual maps, use the check boxes located in the <b>Feature Options</b> portion of the PlasMapper interface. The default settings specify that all feature types in the FSD are shown.
      <tr><td><img SRC="/PlasMapper/image/blueline.gif" ALT="blue line" width="645" height="7"></td></tr>

      <tr><td width=645><b>PlasMapper's restriction site databases:</b><br>
      PlasMapper has two restriction site databases: <a href="/PlasMapper/html/commonEnzyme.html">common</a>, which consists of enzymes routinely used in molecular biology labs, and <a href="/PlasMapper/html/allEnzyme.html">all</a>, which consists of the common enzymes as well as less frequently used enzymes.
      </td></tr>
      <tr><td><img SRC="/PlasMapper/image/blueline.gif" ALT="blue line" width="645" height="7"></td></tr>

      <tr><td width=645><b>Controlling which restriction sites are shown:</b><br>
      PlasMapper provides several options for specifying which restriction sites are labeled on the textual and graphic maps. To label sites found in the <b>common</b> restriction site database, use the radio buttons listed under <b>Use common enzyme set to annotate sequence</b>. To label sites found in the <b>all</b> restriction site database, use the radio buttons listed under <b>Use all enzyme set to annotate sequence</b>. Each radio button set consists of the following:
      <ul>
      <li>
      <b>Unique sites</b> - restriction sites are labeled if they are found exactly once in the plasmid sequence.
      </li>
      <li>
      <b>Unique sites from >= 4 cutters</b> - restriction sites are labeled if they are found exactly once in the plasmid sequence AND they correspond to an enzyme that recognizes a sequence greater than or equal to 4 bases in length.
      </li>
      <li>
      <b>Unique sites from >= 6 cutters</b> - restriction sites are labeled if they are found exactly once in the plasmid sequence AND they correspond to an enzyme that recognizes a sequence greater than or equal to 6 bases in length.
      </li>
      <li>
      <b>All sites</b> - all restriction sites are labeled. The labeled sites may or may not be unique.
      </li>
      <b>All sites from >= 4 cutters</b> - restriction sites are labeled if they correspond to an enzyme that recognizes a sequence greater than or equal to 4 bases in length. The labeled sites may or may not be unique.
      </li>
      <li>
      <b>All sites from >= 6 cutters</b> - restriction sites are labeled if they correspond to an enzyme that recognizes a sequence greater than or equal to 6 bases in length. The labeled sites may or may not be unique.
      </li>
      </ul>

      </td></tr>
      <tr><td><img SRC="/PlasMapper/image/blueline.gif" ALT="blue line" width="645" height="7"></td></tr>

      <tr><td width=645><b>Open reading frame identification:</b><br>
      PlasMapper defines open reading frames (ORFs) as stretches of 200 or more codons starting with an ATG (or the start of the sequence) and containing no terminators (TAA, TAG, TGA). To view ORFs shorter than 200 codons in length, enter a smaller cutoff value in the <b>Display ORFs</b> text area. To prevent maps from being cluttered with too many ORFs, enter a larger value in the <b>Display ORFs</b> text area. ORFs can be hidden for a particular strand of the DNA sequence, by unchecking the <b>Forward Strand</b> or <b>Reverse Strand</b> check boxes.
      </td></tr>
      <tr><td><img SRC="/PlasMapper/image/blueline.gif" ALT="blue line" width="645" height="7"></td></tr>

      <tr><td width=645><b>User defined features:</b><br>
      Features can be manually added to the graphical and textual maps generated by PlasMapper using the <b>User Defined Features</b> section of the PlasMapper interface. This capability has been included so that more specialized features can be labeled, such as species-specific promoters, species-specific terminators, and inserted genes. Up to four features can be defined, by specifying the following:
      <ul>
      <li>
      <b>Name</b> - the label that will be used to mark the feature.
      </li>
      <li>
      <b>Category</b> - the feature type. The feature type determines the default color and shape (arrow or arc) of the feature.
      </li>
      <li>
      <b>Start</b> - the first base in the feature. The start should be less than the stop regardless of the strand of the feature, except when the feature crosses the start/end boundary of the plasmid.
      </li>
      <li>
      <b>Stop</b> - the last base in the feature. The stop should be greater than the stop regardless of the strand of the feature, except when the feature crosses the start/end boundary of the plasmid.
      </li>
      <li>
      <b>Strand</b> - the strand of the feature. Forward strand features are drawn on the outside of the plasmid backbone, while reverse strand features are drawn on the inside of the plasmid backbone.
      </li>
      </ul>
      </td></tr>
      <tr><td><img SRC="/PlasMapper/image/blueline.gif" ALT="blue line" width="645" height="7"></td></tr>

      <tr><td width=645><b>Abbreviations used by PlasMapper:</b><br>
      The feature labels generated by PlasMapper consist of a feature name, followed by an abbreviated feature class designation. For example, the label <b>T3 prom</b> is used to mark the T3 promoter, with <b>T3</b> being the feature name, and <b>promoter</b> being the feature class. Below are the feature classes and abbreviations used by PlasMapper.
      <ul>     
      <li>
      <b>origin</b> - origin of replication.
      </li> 
      <li>
      <b>other</b> - other gene.
      </li> 
      <li>
      <b>prom</b> - promoter.
      </li> 
      <li>
      <b>rf</b> - reading frame.
      </li>  
      <li>
      <b>reg</b> - regulatory sequence.
      </li> 
      <li>
      <b>reporter</b> - reporter gene.
      </li>
      <li>
      <b>marker</b> - selectable marker.
      </li>
      <li>
      <b>tag</b> - tag / epitope.
      </li>
      <li>
      <b>term</b> - terminator.
      </li>
      </ul>
      </td></tr>
      <tr><td><img SRC="/PlasMapper/image/blueline.gif" ALT="blue line" width="645" height="7"></td></tr>
      <tr><td><br><br></td></tr>
      <tr><td><h2>Graphical output options:</h2></td></tr>
      <tr><td><img SRC="/PlasMapper/image/blueline.gif" ALT="blue line" width="645" height="7"></td></tr>

      <tr><td width=645><b>Rendering options:</b><br>
      Several options are available for controlling how PlasMapper generates graphical maps (these options do not affect textual maps):
      <ul>
      <li>
      <b>Color Scheme</b> - <b>Regular</b> or <b>Inverse</b>. <b>Regular</b> generates a map with a white background, while <b>inverse</b> generates a map with a black background. The colors of the other map components are automatically adjusted to suit the background color.
      </li>
      <li>
      <b>Shading</b> - <b>On</b> or <b>Off</b>. When <b>Shading</b> is set to <b>On</b>, map elements are drawn with highlights and shadows. When <b>Shading</b> is set to <b>Off</b>, map elements are drawn without highlights and shadows. Note that shading increases the size of SVG and SVGZ files.
      </li>
      <li>
      <b>Label Colors</b> - <b>On</b> or <b>Off</b>. When <b>Label Colors</b> is set to <b>On</b>, labels are colored to match the feature they represent. When set to <b>Off</b>, labels are colored black or white, depending on the <b>Color Scheme</b>.
      </li>
      <li>
      <b>Labels</b> - <b>On</b> or <b>Off</b>. When <b>Labels</b> is set to <b>On</b>, labels are drawn on the map. When set to <b>Off</b>, labels are not drawn.
      </li>
      <li>
      <b>Inner Labels</b> - <b>On</b> or <b>Off</b>. When <b>Inner Labels</b> is set to <b>On</b>, the PlasMapper rendering engine attempts to place some of the labels on the inside of the backbone circle. Any labels that do not fit on the inside are automatically moved to the outside. When set to <b>Off</b>, all labels are drawn on the outside of the backbone circle.
      </li>
      <li>
      <b>Legend</b> - <b>On</b> or <b>Off</b>. When <b>Legend</b> is set to <b>On</b>, a legend showing the color used for each feature class is displayed. When set to <b>Off</b>, no legend is displayed.
      </li>
      <li>
      <b>Arrows</b> - <b>On</b> or <b>Off</b>. When <b>Arrows</b> is set to <b>On</b>, certain features (promoters, reporter genes, selectable markers, and other genes) are drawn as arrows. When set to <b>Off</b>, all features are drawn as arcs.
      </li>
      <li>
      <b>Tick Marks</b> - <b>On</b> or <b>Off</b>. When <b>Tick Marks</b> is set to <b>On</b>, tick marks are drawn on the map. When set to <b>Off</b>, tick marks are not drawn.
      </li>
      <li>
      <b>Title</b> - text entered in the <b>Title</b> text box is placed in the center of the plasmid backbone.
      </li>
      <li>
      <b>Display Comment</b> - text entered in the <b>Display Comment</b> text box is placed along the bottom of the plasmid map.
      </li>
      <li>
      <b>Image Format</b> - <b>PNG</b> (Portable Network Graphics), <b>JPG / JPEG</b> (Joint Photographic Experts Group), <b>SVG</b> (Scalable Vector Graphics), or <b>SVGZ</b> (compressed SVG). <b>PNG</b> and <b>JPG / JPEG</b> images can be viewed in current web browsers. <b>SVG</b> and <b>SVGZ</b> images can be viewed in current web browsers, using the <a href="http://www.adobe.com/svg/viewer/install/main.html">Adobe SVG plugin</a>. <b>SVG</b> and <b>SVGZ</b> images can also be viewed and edited using vector graphics software, such as Adobe Illustrator.
      </li>
      <li>
      <b>Image Size</b> - the width and height of the graphical map.
      </li>
      <li>
      <b>Backbone Thickness</b> - the thickness of the line used to draw the backbone circle.
      </li>
      <li>
      <b>Arc Thickness</b> - the thickness of the line used to draw the feature arcs and arrows.
      </li>
      </ul>
      </td></tr>

      <tr><td><img SRC="/PlasMapper/image/blueline.gif" ALT="blue line" width="645" height="7"></td></tr>
      <tr><td><br><br></td></tr>
      <tr><td><h2>Frequently Asked Questions (FAQ):</h2></td></tr>
      <tr><td><img SRC="/PlasMapper/image/blueline.gif" ALT="blue line" width="645" height="7"></td></tr>
      <tr><td width=645>
      <ul>
      <li>
      <b>What is the difference between the SVG and SVGZ formats?</b> - SVGZ is simply SVG that has been compressed using the gzip algorithm. The compression produces a smaller file that can be downloaded more quickly. The <a href="http://www.adobe.com/svg/viewer/install/main.html">Adobe SVG plugin</a> converts the SVGZ into SVG automatically.
      </li>
      <li>
      <b>How do I save SVG plasmid maps?</b> - SVG PlasMapper maps are viewed and saved using the <a href="http://www.adobe.com/svg/viewer/install/main.html">Adobe SVG plugin</a>. When viewing a map using the plugin, you can access a menu of controls by right-clicking on the map (if you have a two-button mouse), or by pressing and holding Ctrl while clicking (if you have a single-button mouse). From the menu that appears, select <b>Save SVG</b>. Some versions of the plugin allow you to save the SVG as a .svg file or a .svgz file (compressed svg). On some platforms, the <b>Save SVG</b> command does not function. To get around this limitation, select <b>View Source</b> from the plugin menu (the menu that contains the <b>Save SVG</b> command). Copy and paste the source text into a text editor, and save the file as a plain text file with a .svg extension.
      </li>
      <li>
      <b>How do I edit SVG maps?</b> - The SVG maps generated by PlasMapper can be edited in a variety of vector graphics software packages. For example, Adobe Illustrator 10 and CorelDRAW Graphics Suite 12 support SVG import and editing. To edit an SVG map, first save the map as a .svg or .svgz file, as described above. Some vector graphics packages allow .svg and .svgz files to be opened directly, while others provide an import command for adding the graphics described in the .svg file or .svgz file to a blank canvas. Consult the documentation included with software you are using for specific details on how to open or import .svg and .svgz files.
      </li>
      <li>
      <b>How do I transfer the text output to another program?</b> - Text PlasMapper maps can be copied and pasted into other applications. If the text does not appear to retain the proper layout after transfer, try changing the font to a smaller size, and change the font to a fixed-width font (Courier for example). If blank spaces are removed when the text is pasted, try using a different paste command in the application receiving the text (use <b>Paste Special...</b> for example). Alternatively, use your web browser's <b>View Source</b> command to display the text map source code, and copy and paste the map portion of the source code.
      </li>
      </ul>
      </td></tr>

      <tr><td><img SRC="/PlasMapper/image/blueline.gif" ALT="blue line" width="645" height="7"></td></tr>
      <tr><td><br><br></td></tr>
      <tr><td><h2>Additional information:</h2></td></tr>
      <tr><td><img SRC="/PlasMapper/image/blueline.gif" ALT="blue line" width="645" height="7"></td></tr>
      <tr><td width=645>
      <ul>
      <li>
      <b>Adobe SVG Zone</b> - The <a href="http://www.adobe.com/svg/main.html">Adobe SVG Zone</a> contains information about the Scalable Vector Graphics (SVG) format.
      </li>
      <li>
      <b>Batik SVG Toolkit</b> - PlasMapper uses the <a href="http://xml.apache.org/batik/">Batik SVG Toolkit</a> to generate SVG images.
      </li>
      <li>
      <b>The Feature Site Database (FSD)</b> - <a href="/PlasMapper/html/feature.html">View</a> the database of features used by PlasMapper
      </li>
      <li>
      <b>Restriction enzyme databases</b> - View the <a href="/PlasMapper/html/commonEnzyme.html">common</a> and <a href="/PlasMapper/html/allEnzyme.html">all</a> restriction site lists used by PlasMapper. These sites were obtained from <a href="http://rebase.neb.com/rebase/rebase.html">REBASE</a>.
      </li>
      </ul>
      </td></tr>

 <tr>
    <td><img SRC="/PlasMapper/image/blueline.gif" ALT="blue line" width="645" height="7"><p>
	Problems? Questions? Suggestions? Please contact
      <a href="mailto:xdong@redpoll.pharmacy.ualberta.ca">Xiaoli Dong</a>, <a href="mailto:stothard@ualberta.ca">Paul Stothard</a> or <a HREF="mailto:david.wishart@ualberta.ca">David Wishart</a></td>
  </tr>
<tr><td>Funding for this project was provided by PENCE Inc and Genome Prairie</td></tr>
</td></tr></table>
</body>

</html>
